src/main/java/com/sekwah/advancedportals/bukkit/listeners/Listeners.java

```diff
@@ -137,6 +142,65 @@ public void checkTriggerLocations(Player player, boolean useDelayed, Location...
         }
     }
 
+    class CooldownDataRemovalTask implements Runnable {
+
+        private int removed;
+
+        @Override
+        public void run() {
+            boolean canRemove = true;
+            while (canRemove) {
+                canRemove = Portal.cooldown.entrySet().removeIf(e -> {
+                    HashMap<String, Long> cds = e.getValue();
+                    if (cds == null) {
+                        removed++;
+                        return true;
+                    } else {
+                        cds.entrySet().removeIf(entry -> shouldRemovePortalCooldown(entry));
+                        if (cds.isEmpty()) {
+                            removed++;
+                            return true;
+                        }
+                    }
+                    return false;
+                });
+            }
+
+            // Make sure maps are never too big than they need to be
+            if (removed > 16) {
```

Comment:
I remove there all cooldowns that expired so are not used, then resize hashmaps so it can shrink to more proper size